swarm/src/lib.rs

```diff
@@ -346,37 +381,43 @@ where TBehaviour: NetworkBehaviour<ProtocolsHandler = THandler>,
         // across a `Deref`.
         let this = &mut *self;
 
-        loop {
+        'poll: loop {
             let mut network_not_ready = false;
 
             match this.network.poll(cx) {
                 Poll::Pending => network_not_ready = true,
-                Poll::Ready(NetworkEvent::NodeEvent { conn_info, event }) => {
-                    this.behaviour.inject_node_event(conn_info.peer_id().clone(), event);
+                Poll::Ready(NetworkEvent::ConnectionEvent { connection, event }) => {
+                    let peer = connection.peer_id().clone();
+                    let connection = connection.id();
+                    this.behaviour.inject_event(peer, connection, event);
                 },
-                Poll::Ready(NetworkEvent::Connected { conn_info, endpoint }) => {
-                    if this.banned_peers.contains(conn_info.peer_id()) {
-                        this.network.peer(conn_info.peer_id().clone())
+                Poll::Ready(NetworkEvent::ConnectionEstablished { connection, num_established }) => {
+                    let peer = connection.peer_id().clone();
+                    if this.banned_peers.contains(&peer) {
+                        this.network.peer(peer)
                             .into_connected()
                             .expect("the Network just notified us that we were connected; QED")
-                            .close();
-                    } else {
-                        this.behaviour.inject_connected(conn_info.peer_id().clone(), endpoint);
-                        return Poll::Ready(SwarmEvent::Connected(conn_info.peer_id().clone()));
+                            .disconnect();
+                    } else if num_established == 1 {
+                        let endpoint = connection.endpoint().clone();
+                        this.behaviour.inject_connected(peer.clone(), endpoint);
+                        return Poll::Ready(SwarmEvent::Connected(peer));
                     }
                 },
-                Poll::Ready(NetworkEvent::NodeClosed { conn_info, endpoint, error }) => {
-                    log::trace!("Connection {:?} with endpoint {:?} closed by {:?}",
-                                conn_info, endpoint, error);
-                    this.behaviour.inject_disconnected(conn_info.peer_id(), endpoint);
-                    return Poll::Ready(SwarmEvent::Disconnected(conn_info.peer_id().clone()));
-                },
-                Poll::Ready(NetworkEvent::Replaced { new_info, closed_endpoint, endpoint, .. }) => {
-                    this.behaviour.inject_replaced(new_info.peer_id().clone(), closed_endpoint, endpoint);
+                Poll::Ready(NetworkEvent::ConnectionError { connected, error, num_established }) => {
+                    log::error!("Connection {:?} closed by {:?}", connected, error);
```

Comment:
`error!` should be reserved for exceptional issues, not when a connection closes, which happens all the time.

Reply:
Absolutely, sorry, this is just a left-over from some testing sessions with substrate/polkadot. It should not have been committed. I will change it back to debug level shortly.